packages/client/src/sync/fetcher/fetcher.ts

```diff
@@ -330,19 +336,21 @@ export abstract class Fetcher<JobTask, JobResult, StorageItem> extends Readable
       }
       return false
     }
-    if (this._readableState!.length > this.maxQueue) {
+    const jobStr = this.jobStr(job)
+    if (this._readableState === undefined || this._readableState!.length > this.maxQueue) {
       this.debug(
         `Readable state length=${this._readableState!.length} exceeds max queue size=${
           this.maxQueue
-        }, skip next job execution.`
+        }, skip job ${jobStr} execution.`
       )
       return false
     }
-    if (job.index > this.processed + this.maxQueue) {
-      this.debug(`Job index greater than processed + max queue size, skip next job execution.`)
+    if (job.index > this.finished + this.maxQueue) {
+      this.debug(`Job index greater than finished + max queue size, skip job ${jobStr} execution.`)
+      return false
```

Comment:
Ah, and this is worth to mention as well: this switches from using the processed (before writing to BC) to finished jobs for deciding when to skip.

This avoids (to some extend) that there are too big heaps of yet-to-write-to-chain tasks heaping up (like 16 jobs or so still outstanding). Effect is also not too noticeable though, just gradually helps. No strong opinion and if not wanted we can also switch back.

Reply:
I like this.  Makes more sense in context.